src/utils/raw.rs

```diff
@@ -45,11 +45,9 @@ pub fn open_dir(p: &Path) -> std::io::Result<File> {
 #[cfg(not(windows))]
 pub fn open_dir(p: &Path) -> std::io::Result<File> {
     use std::fs::OpenOptions;
-    use std::os::unix::fs::OpenOptionsExt;
 
     let mut options = OpenOptions::new();
     options.read(true);
-    options.custom_flags(libc::O_NOFOLLOW);
```

Comment:
Do we understand why this code was there in the first place? Seems like this might have security implications?

Reply:
I'm pretty sure it does, as this is the low level helper used to open directories; and rustup deals with many paths that are required to be directories... as opposed to something like ~/.rustup/toolchains which could be either a dir or a symlink.

Reply:
I think changing the code that opens toolchains itself - whereever that is - to resolve the link, sanity check (/ would be invalid), and then open the resolved path, or alternatively open the directory itself, would be a good approach.

Reply:
> I'm pretty sure it does, as this is the low level helper used to open directories; and rustup deals with many paths that are required to be directories... as opposed to something like ~/.rustup/toolchains which could be either a dir or a symlink.

I guess so, but I don't currently see this "helper function" being used elsewhere...

> OTOH the only caller of this function is `Toolchain::exists`, and I don't see a particular reason why we can't follow a symlink for this purpose.
_https://github.com/rust-lang/rustup/issues/3737#issuecomment-2016776591_

...  so even a simple docstring would help in this case.

---

In addition, its Windows version currently DOES follow symlinks right?

> the open with FILE_FLAG_BACKUP_SEMANTICS is required to read the link out of a link, compared to linux where a different syscall is used to read links.
_https://github.com/rust-lang/rustup/issues/3737#issuecomment-2023475496_

https://github.com/rust-lang/rustup/blob/0c501d55c54031a992395078d535b6d2574dc693/src/utils/raw.rs#L33-L44

---

If I understand the current situation correctly, the Windows version of `open_dir()` currently follows symlinks, but the Unix version doesn't; however both versions should either follow or not follow symlinks for the sake of consistency..

Reply:
Renamed `open_dir()` as suggested by @rbtcollins.